Enable pools to consider deferred tasks

Usiel opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:


How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

closes: https://github.com/apache/airflow/discussions/21243

This is my attempt at solving the issue raised in the linked discussion. With this change we would introduce a new flag on `Pool`, that allows an admin to decide whether to include tasks in deferred state in the pool open slot calculation or not. By default we would not consider them, which mirrors the current behavior.

As mentioned by various people on the discussion, having this as an option can be quite useful, especially in cases where the system that is "shielded" by the pool is not completely in the control of the developer. This allows us to (mis)use pools as a rate-limiting device even when using the new deferrable operators.

Usiel committed:
Makes pools respect deferrable tasks (with extra setting)

See https://github.com/apache/airflow/discussions/21243

This commit makes pools consider deferred tasks if the `include_deferred` flag is set. By default a pool will not consider deferred tasks as occupied slots, but still show the number of deferred tasks in its stats.

Usiel committed:
Adds UI for include_deferrable field on Pool

Usiel committed:
Adds documentation for deferrable tasks and pools

Usiel committed:
Adds new metric for deferred tasks in pool

Usiel committed:
Ensures PoolSlotsAvailableDep is met for tasks in deferred state

Adds an extra test and the implementation to ensure that the `include_inferred` field in `Pool` is honored for the `PoolSlotsAvailableDep` checks.

uranusjr commented on airflow/migrations/versions/0128_2_7_0_add_include_deferred_column_to_pool.py:
```diff
@@ -0,0 +1,48 @@
+def upgrade():
+    """Apply add include_deferred column to pool"""
+    with op.batch_alter_table("slot_pool") as batch_op:
+        batch_op.add_column(sa.Column("include_deferred", sa.Boolean, default=False))
```
This should be `nullable=False`

Usiel replied:
Fixed: Added the then required `server_default` and also explicitly set the same arg on the `Pool` class.

RNHTTR commented on docs/apache-airflow/authoring-and-scheduling/deferring.rst:
```diff
@@ -20,7 +20,7 @@ Deferrable Operators & Triggers
 Standard :doc:`Operators </core-concepts/operators>` and :doc:`Sensors <../core-concepts/sensors>` take up a full *worker slot* for the entire time they are running, even if they are idle; for example, if you only have 100 worker slots available to run Tasks, and you have 100 DAGs waiting on a Sensor that's currently running but idle, then you *cannot run anything else* - even though your entire  ...
 
-This is where *Deferrable Operators* come in. A deferrable operator is one that is written with the ability to suspend itself and free up the worker when it knows it has to wait, and hand off the job of resuming it to something called a *Trigger*. As a result, while it is suspended (deferred), it is not taking up a worker slot and your cluster will have a lot less resources wasted on idle Operato ...
+This is where *Deferrable Operators* come in. A deferrable operator is one that is written with the ability to suspend itself and free up the worker when it knows it has to wait, and hand off the job of resuming it to something called a *Trigger*. As a result, while it is suspended (deferred), it is not taking up a worker slot and your cluster will have a lot less resources wasted on idle Operato ...
```
Typo in the following sentence ("task" should be singular)

> Note that by default a deferred task~s~ will not use up a pool slot, if you would like them to, you can change this by editing the pool in question.

Usiel replied:
Oops, fixed by making everything plural

uranusjr commented on airflow/migrations/versions/0128_2_7_0_add_include_deferred_column_to_pool.py:
```diff
@@ -0,0 +1,52 @@
+    with op.batch_alter_table("slot_pool") as batch_op:
+        batch_op.add_column(
+            sa.Column(
+                "include_deferred", sa.Boolean, default=False, nullable=False, server_default=sa.false()
```
Do we still need `default`?

Usiel replied:
Removed it in exchange for a change in the `add_default_pool_if_not_exists` function (needed set the flag there as a result).

uranusjr commented on airflow/models/pool.py:
```diff
@@ -161,21 +163,26 @@ def slots_stats(
 
         pool_rows = session.execute(query)
-        for (pool_name, total_slots) in pool_rows:
+        for (pool_name, total_slots, includes_deferred) in pool_rows:
```
Please unify whether you want to use _include_ or _includes_.

Usiel replied:
Done

uranusjr commented on airflow/models/pool.py:
```diff
@@ -224,15 +236,25 @@ def occupied_slots(self, session: Session = NEW_SESSION) -> int:
+            return EXECUTION_STATES | {
+                TaskInstanceState.DEFERRED,
+            }
+        return EXECUTION_STATES
```
Why is this not an instance method so we can simply do `pool.get_occupied_states()` without arguments?

Usiel replied:
Right, that's fixed. Static method was a remnant of a different approach.

uranusjr commented on airflow/models/pool.py:
```diff
@@ -290,6 +312,25 @@ def scheduled_slots(self, session: Session = NEW_SESSION) -> int:
+            session.scalar(
+                select(func.sum(TaskInstance.pool_slots))
+                .filter(TaskInstance.pool == self.pool)
+                .filter(TaskInstance.state == TaskInstanceState.DEFERRED)
```
```suggestion
                .where(TaskInstance.pool == self.pool, TaskInstance.state == TaskInstanceState.DEFERRED)
```

Usiel replied:
Committed with a fix for formatting 👍

uranusjr reviewed (commented):
I wonder if we could add some help text in the pool form to explain what the flag means, it’s not as straightfoward as other fields. I’m not sure how to do that though, probably via some WTForm trick.

Usiel committed:
Adds link for deferred slots in Pool list

Similar to the other states we add a formatter for the deferred_slots column with a link to TIs with said state.

Usiel committed:
Simplifies Pool task filtering

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

Usiel committed:
Sets include_deferred for all Pool constructor calls

The new required column on Pool requires some changes to the API. The CLI command uses an optional flag `--include_deferred`, which keeps backwards compatibility and is in line with other CLI commands when boolean values are involved.

Usiel committed:
Keep deprecated pool API as is

API doesn't change (except additional fields in results). Behavior for the new include_deferred field on create is:
- Insert: `include_deferred` is set to `False`
- Update: `include_deferred` is not changed

Usiel committed:
Establishes backwards compatibility for v1 pool API

This changeset can be reverted once we want to deprecate requests to the Pool v1 API _without_ the required `include_deferred` field.
For now we allow requests without the required field to pass through, but emit a deprecation warning.

pierrejeambrun reviewed (commented):
We should have a default value at least in the schema and maybe also for the model to autofill the value in case it is not provided. Especially for the API to not make a breaking change. (People need to be able to keep updating/creating without providing this extra field).

Also openAPI spec needs to be adjusted. (`v1.yaml`)

Usiel committed:
Adds deferred slots to Pool v1 API

pierrejeambrun reviewed (dismissed):
Prevent accidental merging before the OpenAPI spec is updated

Usiel committed:
Updates OpenAPI spec to reflect changes to Pool

Usiel committed:
Makes include_deferred non-required for v1 API

Usiel committed:
Removes deprecation handling for include_deferred and makes import/export backwards compatible

pierrejeambrun commented on tests/cli/commands/test_pool_command.py:
```diff
@@ -88,19 +100,34 @@ def test_pool_import_invalid_json(self):
+
+            assert self.session.query(Pool).filter(Pool.pool == "foo").first().include_deferred is False
+        finally:
+            os.remove("pools_import.json")
```
You can use the tmpdir fixture provided by pytest to not have to handle manually this tmp file lifecycle

Usiel replied:
Oh nice! Fixed up the remaining tests as well.

`tmp_path` seems to be the new thing replacing tmpdir ([ref](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html#the-tmpdir-and-tmpdir-factory-fixtures)).

pierrejeambrun commented on airflow/api_connexion/endpoints/pool_endpoint.py:
```diff
@@ -155,3 +156,13 @@ def post_pool(*, session: Session = NEW_SESSION) -> APIResponse:
         raise AlreadyExists(detail=f"Pool: {post_body['pool']} already exists")
+
+
+def _handle_missing_include_deferred(fields_diff: set) -> None:
```
This should be removed

pierrejeambrun reviewed (commented):
~~Yes, and we need something similar for the RestAPI. Try creating a Pool without providing the `include_deferred`, it should work. This can be filled in to False by the [schema]
(https://github.com/apache/airflow/tree/main/airflow/api_connexion/schemas)~~

You already added the `load_default`, we should be good thanks.

Usiel committed:
Cleanup: Removes unused function

Usiel committed:
Refactors tests to clean up generated files via tmp_path fixture

uranusjr commented on airflow/api/client/api_client.py:
```diff
@@ -60,12 +60,13 @@ def get_pools(self):
         raise NotImplementedError()
 
-    def create_pool(self, name, slots, description):
+    def create_pool(self, name, slots, description, include_deferred):
```
Not entirely sure, I think this needs a default `False` for backward compatibility.

Usiel replied:
Internally we cover all our bases upstream of the client, unless I'm missing something. The `pool_command` module is the only consumer and in both usages we handle a missing `include_deferred`:
- When importing a pool from JSON we handle the missing flag and fall back to `False` (https://github.com/apache/airflow/pull/32709/files#diff-b16c716393f690332e062e25defef0da5ac27e282553040b4b4dee223f1d169fR131)
- When upserting a pool (coming from `airflow pools set ...`) we have an _optional_ flag for the CLI command (so the default is `False`) (https://github.com/apache/airflow/pull/32709/files#diff-6e938311edf46fbd7341cc3f151407931322af0121acbfca9885b9f3aa0a8565R516-R518)

Is `get_current_api_client()` (`airflow.api.client`) something that is meant to be used to interact with Airflow? If yes, then yes, I agree we should change the signature as you said. I couldn't find much on this topic, [Public Interface of Airflow docs](https://airflow.apache.org/docs/apache-airflow/stable/public-airflow-interface.html) only makes mention of the stable REST API and CLI.

potiuk replied:
This is old, deprecated, experimental API client that we don't use for years. As such I'd argue that we just should not change it at all.

potiuk replied:
Can you remove those changes from "airflow/api/client" packages @Usiel please? I am eyeing to add that change to 2.7.0 if possible, so if there is a way to do it quickly that would be fantastic.

potiuk replied:
I took the liberty to add a fixup to remove the "experimental api" changes and rebase the change on top of latest main. I will try to convince the release manager to include that one in 2.7.0 - it would be a pity to wait till 2.8 for that one as there is clearly a need.

Usiel replied:
> I will try to convince the release manager to include that one in 2.7.0 - it would be a pity to wait till 2.8 for that one as there is clearly a need.

Thank you, that'd be great!

Unfortunately the `pool_command` makes use of `airflow.api.client`. So if we don't want any changes to the API client, we'd also have to undo the changes for the CLI pool commands (specifically the `airflow pool set ...` cmd). I can do that now, if that's what we're going for.

potiuk replied:
Yeah. just realied taht. My bad. Actually - this was fine how it was before. I realized that the CLI command is still using tha API (Historical reason) but it is NOT the experimental API- it was internal API used in the past by the clients and we moved away from it in newer CLIs. There are absolutely no guarantees about this API that we give - so your change was good as it was. 

resetting my change.
